### DISCUSSION TITLE
Issue with auto-orient

### DISCUSSION THREAD

**Discussion by leetaylor** [ImageMagick/ImageMagick#1959] [General] [answered]
Getting some mixed results using auto-orient in a Ruby project;

Using;
ImageMagick version 6.7.8.9, release 15.21.amzn1
RMagick version 2.13.4 (Ruby 2.6.5)

**Raw example photos;**

https://media.idashboard.com.au/media/photos/19811757/600.jpg
Taken on an iPhone, orientation exif is rotated 180° (3)

https://media.idashboard.com.au/media/photos/19812607/600.jpg
Taken on a Sumsung, orientation exif is rotated 90° (6)

**When we put these 2 images through auto_orient, it seems to be able to fix the iPhone photo, but not the Samsung photo:**

http://media-server-v4-staging.us-west-1.elasticbeanstalk.com/media/photos/19811757/600.jpg
Photo is correctly rotated and the orientation exif is now 1

http://media-server-v4-staging.us-west-1.elasticbeanstalk.com/media/test/19812607/600.jpg
Photo doesn't get rotated, but orientation exif is changed to 1 anyway

We're performing auto_orient on its own, _after_ we fetch the original image from S3 and _before_ we go into any image manipulation/resizing/etc. Totally confused why it manages to fix the iPhone photo, but not the Samsung photo. No idea if it is a bug, or it is the Samsung exif, but any help would be really appreciated.

Just to bypass Ruby/RMagick as a potential cause, I ran auto-orient via command line and got the same result on the Samsung photo

`convert "original.jpg" -auto-orient "output.jpg"`

http://media-server-v4-staging.us-west-1.elasticbeanstalk.com/output.jpg

**Comment by fmw42** [comment]
I tried the same command with IM 6.9.11.9 and 7.0.10.9 and the second one does not auto-rotate. But the first one visually (from the preview) looks like it is already properly oriented. I am using Mac Preview.

**Reply by leetaylor** [reply]
I get mixed results on both photos in different browsers, mac vs. pc and when I download them. It's really got me confused.

I.e. Original iPhone photo looks fine on Mac previewer (go figure), fine in Firefox when looking at the image url directly, but then it is upside down when Firefox renders the image inside a page.

Samsung photo is sideways on Mac previewer, sideways in Firefox when looking at the image url directly, but then it is correctly orientated when Firefox renders the image inside a page.

**Reply by fmw42** [reply]
ImageMagick is finding the Orientation from the meta data, because it converts them to corner names in identify -verbose.  I have tried a number of versions of ImageMagick and they all fail. So unless my syntax is wrong (same as yours), perhaps the files are corrupt in some way. Or ImageMagick has been in error for a large number of version. This will need one of the IM developers to look into it.

**Reply by leetaylor** [reply]
Thanks for taking the time to look at it, much appreciated :)

**Comment by snibgo** [comment, marked as answer]
I don't think IM has a problem.

The Samsung photo, 600,jpg, contains a portrait-format image, 450x600 pixels, with the floor at the bottom. So it doesn't need rotating. But the orientation metadata is 6, "rotate 90 clockwise".

Hence, most photo viewers will obey the orientation flag, and display the photo rotated (with the floor on the left side). If we use IM "-auto-rotate", it will rotate in the same way. But we don't want this, because the metadata is wrong.

**Reply by leetaylor** [reply]
I think you might be right.. I just did another review of the Samsung photo;

Displays wrong (floor to the left) on Mac Preview, Firefox direct to image, Chrome direct to image, Chrome rendered in a page.

For some reason it appears correct (floor at bottom) on Firefox when rendered in a page.. which suggests to me that Firefox rendering is possibly in the wrong here and the image is genuinely not correct?

I use an iPhone, sometimes the gyro won't alter the camera orientation and you have to shake the crap of of the phone to finally work out which way is up.. I'm assuming Samsung might be the same where the orientation gets stuck and the user doesn't notice?

**Reply by snibgo** [reply]
> and the image is genuinely not correct?

I repeat: the image itself is fine. It is correct. It has the floor at the bottom. The orientation metadata is incorrect.

I don't know why the metadata is incorrect. DSLRs always (as far as I know) store images in landscape format. Your Samsung photo is stored in portrait format. Is this normal for the Samsung? If so, I would suspect a faulty camera sensor. But if the Samsung, like DSLRs, stores images in landscape format then some software has rotated the image without zeroing the metadata.

**Reply by leetaylor** [reply]
Thanks, snigbo - sorry, totally misinterpreted your original response.

Understand now that the 'raw' photo is with the floor on the bottom, but since the orientation says 90 cw, it renders with the floor to the left. And auto-orient is technically doing the 'right' thing here because the exif says 90 cw.

Thanks heaps for taking the time to look into it for me 🥇 my testing and all the different results was driving me crazy!
